formats/properties/commonMain/src/kotlinx/serialization/properties/Properties.kt

```diff
@@ -0,0 +1,181 @@
+            return map.getValue(tag)
+        }
+
+        override fun decodeTaggedEnum(tag: String, enumDescriptor: SerialDescriptor): Int {
+            return when (val taggedValue = map.getValue(tag)) {
+                is Int -> taggedValue
+                is String -> enumDescriptor.getElementIndex(taggedValue)
+                else -> throw SerializationException("Value of enum entry '$tag' is neither an Int nor a String")
+            }
+        }
+
+        override fun decodeElementIndex(descriptor: SerialDescriptor): Int {
+            while (currentIndex < size) {
+                val name = descriptor.getTag(currentIndex++)
+                if (map.keys.any { it.startsWith(name) }) return currentIndex - 1
+                if (isCollection) {
+                    // if map does not contain key we look for, then indices in collection have ended
+                    break
+                }
+            }
+            return CompositeDecoder.DECODE_DONE
+        }
+    }
+
+    /**
+     * Stores properties from given [value] to a map and returns this map.
+     * `null` values are omitted from the output.
+     */
+    public fun <T> encodeToMap(strategy: SerializationStrategy<T>, value: T): Map<String, Any> {
+        val m = OutMapper()
+        m.encodeSerializableValue(strategy, value)
+        return m.map
+    }
+
+    /**
+     * Loads properties from given [map], assigns them to an object and returns this object.
+     * [T] may contain properties of nullable types; they will be filled by non-null values from the [map], if present.
+     */
+    public fun <T> decodeFromMap(strategy: DeserializationStrategy<T>, map: Map<String, Any>): T {
+        val m = InMapper(map, strategy.descriptor)
+        return m.decodeSerializableValue(strategy)
+    }
+
+    @Deprecated(removedMsg, level = DeprecationLevel.ERROR)
+    public fun <T> storeNullable(strategy: SerializationStrategy<T>, value: T): Map<String, Any?> = noImpl()
+
+    @Deprecated(removedMsg, level = DeprecationLevel.ERROR)
+    public fun <T> loadNullable(strategy: DeserializationStrategy<T>, map: Map<String, Any?>): T = noImpl()
+
+    @Deprecated(removedMsg, level = DeprecationLevel.ERROR)
+    public inline fun <reified T : Any> storeNullable(value: T): Map<String, Any?> =
+        noImpl()
+
+    @Deprecated(removedMsg, level = DeprecationLevel.ERROR)
+    public inline fun <reified T : Any> loadNullable(map: Map<String, Any?>): T = noImpl()
+
+    /**
+     * A [Properties] instance that can be used as default and does not have any [SerializersModule] installed.
+     */
+    public companion object Default : Properties(EmptySerializersModule, null)
+}
+
+private class PropertiesImpl(serializersModule: SerializersModule) : Properties(serializersModule, null)
+
+/**
+ * Create an instance of [Properties] with a given [module].
+ */
+public fun Properties(module: SerializersModule): Properties = PropertiesImpl(module)
+
+/**
+ * A reified version of [encodeToMap].
```

Comment:
Let's refrain from "reified version of X" in the documentation and make it more self-contained. 
E.g. like this: https://github.com/Kotlin/kotlinx.serialization/blob/dev/runtime/commonMain/src/kotlinx/serialization/SerialFormat.kt#L65